common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java

```diff
@@ -0,0 +1,133 @@
+        Camera camera = event.getCamera();
+        Vec3 cameraPos = camera.getPosition();
+        Vector3f positiveX = new Vector3f(1.0f, 0.0f, 0.0f);
+        Vector3f positiveY = new Vector3f(0.0f, 1.0f, 0.0f);
+
+        // apply camera rotation
+        projection.multiply(positiveX.rotationDegrees(camera.getXRot()));
+        projection.multiply(positiveY.rotationDegrees(camera.getYRot() + 180.0F));
+
+        // offset to put text to the center of the block
+        float dx = (float) (location.x + 0.5 - cameraPos.x);
+        float dy = (float) (location.y + 0.5 - cameraPos.y);
+        float dz = (float) (location.z + 0.5 - cameraPos.z);
+
+        double squaredDistance = dx * dx + dy * dy + dz * dz;
+
+        double distance = Math.sqrt(squaredDistance);
+        int maxDistance = McUtils.mc().options.renderDistance * 16;
+
+        this.distanceText = Math.round((float) distance) + "m";
+
+        // move the position to avoid ndc z leak past 1
+        if (distance > maxDistance) {
+            double posScale = maxDistance / distance;
+            dx *= posScale;
+            dz *= posScale;
+        }
+
+        worldToNdc(new Vector3f(dx, dy, dz), projection);
+    }
+
+    @SubscribeEvent
+    public void onRenderGuiPost(RenderEvent.Post event) {
+        if (CompassModel.getCompassLocation().isEmpty() || ndc == null || isOnScreen()) return;
+
+        Window window = event.getWindow();
+
+        float backgroundWidth = FontRenderer.getInstance().getFont().width(distanceText);
+        float backgroundHeight = FontRenderer.getInstance().getFont().lineHeight;
+
+        float displayPositionX = (float) ((ndc.x + 1.0f) / 2.0f) * window.getGuiScaledWidth();
+        float displayPositionY = (float) ((1.0f - ndc.y) / 2.0f) * window.getGuiScaledHeight();
+
+        RenderUtils.drawRect(
+                event.getPoseStack(),
+                CommonColors.BLACK.withAlpha(backgroundOpacity),
+                displayPositionX - (backgroundHeight / 2) - 2,
+                displayPositionY - 2,
+                0,
+                backgroundWidth + 3,
+                backgroundHeight + 2);
+        FontRenderer.getInstance()
+                .renderAlignedTextInBox(
+                        event.getPoseStack(),
+                        distanceText,
+                        displayPositionX - (backgroundHeight / 2),
+                        displayPositionX + backgroundWidth,
+                        displayPositionY,
+                        displayPositionY + backgroundHeight,
+                        textColor,
+                        HorizontalAlignment.Left,
+                        textShadow);
+    }
+
+    // ndc range check, make sure the point is on screen
+    // z check is flipped to countereact matrix transformation in onRenderGuiPost
+    // possible behavior change?
+    private boolean isOnScreen() {
+        return !(ndc.x > 1 || ndc.x < -1 || ndc.y > 1 || ndc.y < -1 || !(ndc.z > 1));
```

Comment:
Bit weird, you could flip the whole thing to remove `!`?

Reply:
the check is a somewhat specific thing, I fear flip the operator might be unintutive